Additional restrictions

Licenser opened the pull request:
Add restriction lints for `panic!`, `unreachable!`, `todo!` and `.expect(...)`

changelog: Add 5 new `restriction` lints: `panic`, `unreachable`, `todo`, `option_expect_used`, `result_expect_used`

llogiq commented:
Since you have `panic!` and `unimplemented!`, you probably also want to catch `todo!` which was recently sent on its way to stabilization.

Licenser commented:
yes! thanks I didn't knew that it makes total sense :D

flip1995 commented on tests/ui/panic_unimplemented.rs:
```diff
@@ -65,4 +77,6 @@ fn main() {
     ok_nomsg();
     ok_escaped();
     unimplemented();
+    unreachable();
```
Since you separated out `panic!`, can you also separate these test cases?

Licenser replied:
I would have left panic in but panic and panic with sensible message clash.

flip1995 replied:
We want to split up the tests as much as possible, see #2038. Since the unreachable, panic, unimplemented and todo tests are pretty small, it may make sense to group them in a `panicking_macros.rs` test file.

bors commented:
:umbrella: The latest upstream changes (presumably #4619) made this pull request unmergeable. Please resolve the merge conflicts.

Licenser commented:
The failures seem to relate to something outside of the PR, and I can't run tests locally any more either because of a dependency failing, is that a known issue?

llogiq commented:
We likely need another rustup. Hang in there, we'll tell you when it's ready to rebase.

flip1995 commented on clippy_lints/src/methods/mod.rs:
````diff
@@ -90,6 +90,63 @@ declare_clippy_lint! {
+    /// ```
+    pub OPTION_EXPECT_USED,
+    restriction,
+    "using `Option.unwrap()`, which should at least get a better message using `expect()`"
````
```suggestion
    "using `Option.expect()`, which might be better handled"
```

flip1995 commented on clippy_lints/src/methods/mod.rs:
```diff
@@ -2044,6 +2129,7 @@ fn lint_ok_expect(cx: &LateContext<'_, '_>, expr: &hir::Expr, ok_args: &[hir::Ex
         if let Some(error_type) = get_error_type(cx, result_type);
         if has_debug_impl(error_type, cx);
 
+
```
```suggestion
```

flip1995 commented on clippy_lints/src/panic_unimplemented.rs:
````diff
@@ -25,6 +25,22 @@ declare_clippy_lint! {
+    /// ```
+    pub PANIC,
+    restriction,
+    "missing parameters in `panic!` calls"
````
copy pasted

Licenser replied:
fixed :)

flip1995 commented on clippy_lints/src/panic_unimplemented.rs:
```diff
@@ -41,7 +57,39 @@ declare_clippy_lint! {
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of `unreachable!`.
+    ///
+    /// **Why is this bad?** This macro can cause cause code to panics
```
```suggestion
    /// **Why is this bad?** This macro can cause code to panic
```

flip1995 commented on tests/ui/methods.rs:
```diff
@@ -1,7 +1,13 @@
+    clippy::pedantic,
+    clippy::option_unwrap_used,
+    clippy::option_expect_used,
+    clippy::result_expect_used
```
Can you put the `expect` tests in their own file? See #2038

Licenser replied:
i can, i'm a bit unsure what the methodology is, from the tests I looked at it usually was "file per lint test" and tried to stick that, i.e. the `expect` tests are in methods as they're in the lint pass for methods, the `unreachable` is in the panic_unimplemented as it's in that lint pass. I couldn't put `panic` in panic_unimplemented since it clashed with the other panic tests there but looking at methods.rs I could add it as a warning.

So should I put all the new lints into a new file or make one file each or combine them all into their lint passes?

flip1995 replied:
"one file per lint" is a goal, that we want to achieve, but we're not really there yet, as you can see in the `methods.rs` test file. (Also "one file per lint" can also mean grouping some of the lints, when it makes sense, e.g. the 4 panic macro `Lint`s) 

For clarification: `Lint != LintPass`. One file per `LintPass` would be too big, since for example the `methods` module uses a single `LintPass` to check X (X=big number, that I don't want to count) `Lint`s.

> So should I put all the new lints into a new file or make one file each or combine them all into their lint passes?

What I would suggest is to put the `expect` lint tests in one file and the panic macro tests in their own file. 

So for the currently implemented lints, you should create 2 test files:

1. `tests/ui/panicking_macros.rs`, which tests:
    - `panic`
    - `unimplemented` (Move this tests out of `panic_unimplemented.rs` into the new file; also maybe rename the `panic_unimplemented.rs` file)
    - `unreachable`
    - `todo`
2. `tests/ui/expect`, which tests:
    - `option_expect_used`
    - `result_expect_used`

---

Bonus points, if you also move the `unwrap` lints to their own file.

flip1995 commented:
Rustup is done and Clippy master builds again.

Licenser commented:
Split out the tests into two now sets `expect_unwrap` that cover `.expct("")` and `.unwrap("")` for `Result` and `Option` - I moved one test from methods in there. And as suggested `panicking_macros`.

flip1995 commented:
Tests look really good now, thanks!

If you want to, we can merge this PR already as it is right now and you can add the remaining 2 lints  in another PR. Only thing left to do here, would be to squash some of the commits.

Licenser commented:
Ja I wanted to ask about the squashing policy anyway :D will do and rebase so we can pull them in.

Licenser commented:
Commits are squashed!

bors commented:
:umbrella: The latest upstream changes (presumably #4683) made this pull request unmergeable. Please resolve the merge conflicts.

flip1995 commented:
Thanks, I like PRs that add 5 new lints at once 🎉 

@bors r+

bors commented:
:pushpin: Commit 7f454d8d06279663a330cfc4c9248056c402acd1 has been approved by `flip1995`

<!-- @bors r=flip1995 7f454d8d06279663a330cfc4c9248056c402acd1 -->
<!-- homu: {"type":"Approved","sha":"7f454d8d06279663a330cfc4c9248056c402acd1","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 7f454d8d06279663a330cfc4c9248056c402acd1 with merge c0b2411f06a417266baaac2f0ea431138da33bf3...
<!-- homu: {"type":"BuildStarted","head_sha":"7f454d8d06279663a330cfc4c9248056c402acd1","merge_sha":"c0b2411f06a417266baaac2f0ea431138da33bf3"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/132514102), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/28204485)
Approved by: flip1995
Pushing c0b2411f06a417266baaac2f0ea431138da33bf3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/28204485","checks-travis":"https://travis-ci.com/rust-lang/rust-clippy/builds/132514102"},"merge_sha":"c0b2411f06a417266baaac2f0ea431138da33bf3"} -->
